### PR TITLE
[tools/model2nnpkg] Support adding connection information

### PR DESCRIPTION
**Pull request by ragmani**
This commit make `model2nnpkg.sh` support adding connection information between models to metadata.

ONE-DCO-1.0-Signed-off-by: ragmani <ragmani0216@gmail.com>

### CODE REVIEW

**Comment by ragmani** [comment]
Related to https://github.sec.samsung.net/STAR/nnfw/issues/12627#issuecomment-1800581

```bash
$ python3 tools/tflitefile_tool/select_operator.py inceptionv3_slim_2016.tflite <( echo "0-94" ) i1.tflite --store-io-info i1.json
Input tensor(s): [247]
Output tensor(s): [165, 173, 175]
Append subgraphs, orginal index :  0 , new index :  0

$ python3 tools/tflitefile_tool/select_operator.py inceptionv3_slim_2016.tflite <( echo "95-108" ) i2.tflite --store-io-info i2.json
Input tensor(s): [0, 1, 2]
Output tensor(s): [25]
Append subgraphs, orginal index :  0 , new index :  0

$ python3 tools/tflitefile_tool/select_operator.py inceptionv3_slim_2016.tflite <( echo "109-113" ) i3.tflite --store-io-info i3.json
Input tensor(s): [0]
Output tensor(s): [1, 9]
Append subgraphs, orginal index :  0 , new index :  0

$ python3 tools/tflitefile_tool/select_operator.py inceptionv3_slim_2016.tflite <( echo "114-120" ) i4.tflite --store-io-info i4.json
Input tensor(s): [0]
Output tensor(s): [9, 11]
Append subgraphs, orginal index :  0 , new index :  0

$ python3 tools/tflitefile_tool/select_operator.py inceptionv3_slim_2016.tflite <( echo "121-124" ) i5.tflite --store-io-info i5.json
Input tensor(s): [0, 1, 2, 3]
Output tensor(s): [9]
Append subgraphs, orginal index :  0 , new index :  0



$ ./tools/nnpackage_tool/model2nnpkg/model2nnpkg.sh -m i1.tflite i2.tflite i3.tflite i4.tflite i5.tflite -p temp/iv -i i1.json i2.json i3.json i4.json i5.json
$ cat temp/iv/metadata/MANIFEST 
{
  "major-version" : "1",
  "minor-version" : "2",
  "patch-version" : "0",
  "configs"     : [  ],
  "models"      : [ "i1.tflite", "i2.tflite", "i3.tflite", "i4.tflite", "i5.tflite" ],
  "model-types" : [ "tflite", "tflite", "tflite", "tflite", "tflite" ],
  "pkg-inputs"    : [ "0:0:0","0:0:1","0:0:2","1:0:0","1:0:0","2:0:0","2:0:1","3:0:0","3:0:1" ],
  "pkg-outputs"   : [ "1:0:0","1:0:1","1:0:2","2:0:0","3:0:0","4:0:0","4:0:1","4:0:2","4:0:3" ],
  "model-connect" : [ { "from" : "3:0:1", "to" : [ "4:0:3" ] },{ "from" : "2:0:1", "to" : [ "4:0:1" ] },{ "from" : "0:0:0", "to" : [ "1:0:0" ] },{ "from" : "0:0:2", "to" : [ "1:0:2" ] },{ "from" : "2:0:0", "to" : [ "4:0:0" ] },{ "from" : "1:0:0", "to" : [ "2:0:0","3:0:0" ] },{ "from" : "3:0:0", "to" : [ "4:0:2" ] },{ "from" : "0:0:1", "to" : [ "1:0:1" ] } ]
}
```

File: tools/nnpackage_tool/model2nnpkg/model2nnpkg.sh [hunk 1]
```diff
@@ -111,6 +129,68 @@ do
...
+        continue
+      fi
+
+      output_pos=0
+
+      for org_output in $(echo "${model_outputs}" | jq '.org[]'); do
+        if [[ $org_input == $org_output ]]; then
+          to=$(echo "$in_model:0:$input_pos" | jq -R)
+          from=$(echo "$out_model:0:$output_pos" | jq -R)
+          pkg_inputs_str+="$from,"
```
Comment (ragmani) [inline tools/nnpackage_tool/model2nnpkg/model2nnpkg.sh:168]: `from` is an output of a separate model, is it correct to add it to `pkg-inputs`? I'm a little confused.

Comment (glistening) [reply tools/nnpackage_tool/model2nnpkg/model2nnpkg.sh:168]: No. `pkg-inputs`is the user input. The internal connection between models should not be in `pkg-inputs`.

**Comment by glistening** [comment]
>   "pkg-inputs"    : [ "0:0:0","0:0:1","0:0:2","1:0:0","1:0:0","2:0:0","2:0:1","3:0:0","3:0:1" ],

- I think "0:0:0" is the only `pkg-inputs`.
- Additionally, "1:0:0","1:0:0" — duplicated inputs. (`1:0:0` itself is not pkg-inputs.)

File: tools/nnpackage_tool/model2nnpkg/model2nnpkg.sh (hunk 1 above)
Comment (ragmani) [reply tools/nnpackage_tool/model2nnpkg/model2nnpkg.sh:168]: Oh, I was misunderstanding. I will fix it.

**Comment by ragmani** [comment]
```bash
$ ./tools/nnpackage_tool/model2nnpkg/model2nnpkg.sh -m i1.tflite i2.tflite i3.tflite i4.tflite i5.tflite -p iv3 -i i1.json i2.json i3.json i4.json i5.json
model2nnpkg.sh: Generating nnpackage iv3 in .
$ cat iv3/metadata/MANIFEST 
{
  "major-version" : "1",
  "minor-version" : "2",
  "patch-version" : "0",
  "configs"     : [  ],
  "models"      : [ "i1.tflite", "i2.tflite", "i3.tflite", "i4.tflite", "i5.tflite" ],
  "model-types" : [ "tflite", "tflite", "tflite", "tflite", "tflite" ],
  "pkg-inputs"    : [ 0:0:0 ],
  "pkg-outputs"   : [ 4:0:0 ],
  "model-connect" : [ { "from" : "1:0:0", "to" : [ "2:0:0","3:0:0" ] },{ "from" : "0:0:1", "to" : [ "1:0:1" ] },{ "from" : "3:0:0", "to" : [ "4:0:2" ] },{ "from" : "2:0:1", "to" : [ "4:0:1" ] },{ "from" : "3:0:1", "to" : [ "4:0:3" ] },{ "from" : "2:0:0", "to" : [ "4:0:0" ] },{ "from" : "0:0:2", "to" : [ "1:0:2" ] },{ "from" : "0:0:0", "to" : [ "1:0:0" ] } ]
}
```

**Comment by ragmani** [comment]
@glistening PTAL

**Comment by glistening** [comment]
@ragmani I think 3 outputs from 1st model connected to 1st (the only) input of 2nd model. Is it right?

File: tools/nnpackage_tool/model2nnpkg/README.md [hunk 2]
```diff
@@ -4,20 +4,34 @@
 
 It takes `modelfile` as input and generates `nnpackage`.
 
+## Prerequisite
+
+Install jq
+```
+$ sudo apt-get install jq
+```
+
```
Comment (hseok-oh) [inline tools/nnpackage_tool/model2nnpkg/README.md:13]: `model2nnpkg.sh` is used on compiler's command `one-pack`.
This dependency make additional requirement to compiler.

**Comment by hseok-oh** [comment]
@seanshpark @mhs4670go Is it OK to change `model2nnpkg` bash script as python script?

File: tools/nnpackage_tool/model2nnpkg/README.md (hunk 2 above)
Comment (ragmani) [reply tools/nnpackage_tool/model2nnpkg/README.md:13]: It's too difficult to parse json without `jq` in `model2nnpkg`.
If I cannot use `jq` in `model2nnpkg`, it looks like I'll have to reconfigure `MENIFEST` file outside without modifying this script.

**Comment by mhs4670go** [comment]
@hseok-oh Seems okay to me. `one-compiler` toolchain only cares about the command line interface.

**Comment by seanshpark** [comment]
> Is it OK to change model2nnpkg bash script as python script?

If it is not version dependent or requires any packages then it is OK.

**Comment by hseok-oh** [comment]
Let's introduce python script `model2nnpkg` as I told on offline.

**Comment by ragmani** [comment]
This feature is going to be implemented in newly introduced python script.

**Pull request closed without merging**